tencentcloud/resource_tc_cbs_snapshot.go

```
@@ -0,0 +1,287 @@
+		Message  string
+		CodeDesc string
+	}
+	err = json.Unmarshal([]byte(response), &jsonresp)
+	if err != nil {
+		return err
+	}
+	if jsonresp.Code != 0 {
+		return fmt.Errorf(
+			"ModifySnapshot error, code:%v, message: %v, codeDesc: %v.",
+			jsonresp.Code,
+			jsonresp.Message,
+			jsonresp.CodeDesc,
+		)
+	}
+
+	log.Printf("[DEBUG] ModifySnapshot, new snapshotName: %#v.", snapshotName)
+	return nil
+}
+
+func deleteSnapshot(snapshotId string, client *client.Client) *resource.RetryError {
+	params := map[string]string{
+		"Action":        "DeleteSnapshot",
+		"snapshotIds.0": snapshotId,
+	}
+
+	response, err := client.SendRequest("snapshot", params)
+	if err != nil {
+		return resource.NonRetryableError(err)
+	}
+
+	var jsonresp struct {
+		Code     int
+		Message  string
+		CodeDesc string
+		Detail   map[string]struct {
+			Msg  string
+			Code int
+		}
+	}
+	err = json.Unmarshal([]byte(response), &jsonresp)
+	if err != nil {
+		return resource.NonRetryableError(err)
+	}
+	if jsonresp.Code != 0 {
+		return resource.NonRetryableError(fmt.Errorf(
+			"DeleteSnapshot error, code:%v, message: %v, codeDesc: %v.",
+			jsonresp.Code,
+			jsonresp.Message,
+			jsonresp.CodeDesc,
+		))
+	}
+	code := jsonresp.Detail[snapshotId].Code
+	msg := jsonresp.Detail[snapshotId].Msg
+	if code == 16003 || code == 0 {
+		return nil
+	} else if code == 16004 || code == 16033 {
+		return resource.RetryableError(fmt.Errorf("snapshot status error, please retry later"))
+	} else {
+		return resource.NonRetryableError(fmt.Errorf("DeleteSnapshot failed, inner code:%v, message: %v", code, msg))
+	}
+	return nil
+}
+
+func waitingSnapshotReady(snapshotId string, client *client.Client) error {
+	for i := 0; i < 200; i++ {
+		snapshotInfo, _, err := describeSnapshot(snapshotId, client)
+		if err != nil {
+			return err
+		}
+		if snapshotInfo.SnapshotStatus == "creating" {
+			log.Printf("[DEBUG] waiting snapshot ready")
+			time.Sleep(time.Second * 3)
+			continue
+		}
+		break
```

error will be swallowed, if the creating take super loooooong.

---

also this could be refactor with tf's Retry function with specified timeout para.